### ISSUE TITLE
why isPlainObject?

### ISSUE THREAD

**Issue by Santinell** [apigee-127/swagger-tools#284]
Why you use isPlainObject in handlerCacheFromDir?

I want use complex Objects for controllers. In ideal - ES6 classes.
for ES5 Object methods can be used _.each
for ES6 classes is needed workaround.

**Comment by whitlockjc** [comment]
I use `isPlainObject` because the API expects a map of key/value pairs.  I can relax things a bit to let you pass in a class instance but in 1.5 years, this is the first request so I'm not sure how useful it would be other than to you.

**Comment by Santinell** [comment]
Hmm... i think i can generate map and use it in options.controllers

**Comment by whitlockjc** [comment]
I'm closing this for now but if you get something in mind for class-based stuff, let me know.  Also, PRs are appreciated.  :)

**Comment by pixelshaded** [comment]
+1 to this. We use typescript to generate our controllers and were exporting instances of a class which is not compatible with isPlainObject. Means we can't create controllers that inherit from other controllers. isPlainObject is very limiting.

**Comment by whitlockjc** [comment]
I don't mind changing `isPlainObject` to `isObject` as it really doesn't need to be that specific.

**Comment by pixelshaded** [comment]
yes but isObject allows arrays, functions, regexes, new Number(0), and new String('') which I don't think you accept.

**Comment by pixelshaded** [comment]
the work around for us may be to us _.toPlainObject on the instance before exporting.

**Comment by whitlockjc** [comment]
Well, that's why I did this in the first place.  It use to be `isObject`.  I'll figure something out.

**Comment by pixelshaded** [comment]
I tried _.toPlainObject and it destroyed the references to 'this'. This project was actually passing in a map object with instances before. I was refactoring to the use the folder convention instead and thats when stuff stopped working.

**Comment by pixelshaded** [comment]
could always brute force it:

if (_.isObject(controller) && !_.isArray(controller) && !_.isFunction(controller) && !_.isRegExp(controller))

**Comment by whitlockjc** [comment]
That's ugly but if that's what I have to do because there is no other option, I don't mind.

**Comment by pixelshaded** [comment]
I've been looking around for some solutions. This may work:

```javascript
Object.prototype.toString.call(controller) === '[ object Object]'
```

**Comment by pixelshaded** [comment]
Actually I was slightly wrong on it working before. Before we had our own middleware imlementation: we would get the operation and controller info from the meta data middleware and use that to check for the existence of said method on controller and then execute it. It seems that swagger-router looks at the object and scans every property to put it in a handler cache. This means that every property on the controller must be a function. Since our controllers are classes, they may have some members that aren't functions. This is causing throws on start up. I may just need to revert back to our custom router. We were only using swagger-router for stubs.

**Comment by whitlockjc** [comment]
Are you getting errors or warnings?  The reason I ask is if I load your modules, I will skip non-function exports gracefully and give you debug output: https://github.com/apigee-127/swagger-tools/blob/master/middleware/swagger-router.js#L96  If you're providing me with the controller map, the handler itself must be a function and I will error in that case: https://github.com/apigee-127/swagger-tools/blob/master/middleware/swagger-router.js#L369

**Comment by pixelshaded** [comment]
Providing the controller map doesn't work for me because the _.each is going over every property on the controller, throwing if they aren't functions rather than ignoring them. I would argue that non function properties should be ignored since they obviously couldn't be handlers.

Since we are using complex classes, the controller may have a property on it that holds a logger for instance (that is used in other methods). The existence of this logger will cause swagger-router to throw.

Here is our implementation of a swagger router:
```typescript
import express = require('express');
import swaggerTools = require('swagger-tools');
import requireDir = require('require-dir');
import loggerProvider from '../init/LoggerProvider';
import * as _ from 'lodash';

export default (controllersPath: string): express.RequestHandler => {

  const controllerMap: { [controllerName: string]: any } = requireDir(controllersPath);

  let requestHandler: express.RequestHandler = (req: swaggerTools.IRequest, res: express.Response, next: express.NextFunction) => {
    if (!req.swagger ||
      !req.swagger.path ||
      !req.swagger.path['x-swagger-router-controller'] ||
      !req.swagger.operation ||
      !req.swagger.operation.operationId) {
      next();
      return;
    }

    let controllerName: string = req.swagger.path['x-swagger-router-controller'];
    let methodName: string = req.swagger.operation.operationId;

    if (!controllerMap[controllerName]) {
      loggerProvider.logger.warn(`${controllerName} did not exist in ${controllersPath}.`);
      next();
      return;
    }

    if (!controllerMap[controllerName][methodName]) {
      loggerProvider.logger.warn(`${methodName} operation did not exist on swagger controller ${controllerName}.`);
      next();
      return;
    } else if (!_.isFunction(controllerMap[controllerName][methodName])) {
      loggerProvider.logger.warn(`${methodName} operation on swagger controller ${controllerName} was not a function.`);
      next();
      return;
    }

    controllerMap[controllerName][methodName](req, res, next);
  };

  return requestHandler;
};
```

**Comment by whitlockjc** [comment]
I think I better understand the issue now and I've come to the conclusion that it's working as intended.  `options.controllers` is documented such that when it is an object, the key is the routing key _(`[{ControllerName}_]{HandlerFunctionName}`)_ and the value is the function that handles the requests, not a class or a controller or a module.  The problem is that `require-dir` will not give you a mapping that adheres to the contract.  It shouldn't take much effort to take the output from `require-dir` to create a valid controller map.

The whole purpose behind allowing `options.controllers` to be an object provided by the user was to allow the user to manually generate the controller map for cases where we couldn't do it for them.  This also allows the user to use ES6/TypeScript/... without swagger-tools having to explicitly support these technologies.  It also meant that swagger-tools didn't have to implement IoC or anything else but it does mean that the server author has to generate a valid `options.controllers`.

Here is an example: https://github.com/apigee-127/swagger-tools/blob/master/test/2.0/test-middleware-swagger-router.js#L121

**Comment by pixelshaded** [comment]
Looks like I had made the assumption that it was controllerName as the key and the value was the instance with all the methods (operation ids) as properties.

Does passing in a folder iterate over every property on the module and validate it is a function as well?

**Comment by whitlockjc** [comment]
If you pass in a folder, it will attempt to `require` each module in that folder and iterate over each export.  But...if the exported value is not a function, it is logged and skipped.

**Comment by pixelshaded** [comment]
Cool. What were your thoughts on using Object.prototype.string instead of lodash's isPlainObject?

**Comment by whitlockjc** [comment]
I don't mind doing that.

**Comment by pixelshaded** [comment]
make it so :)

**Comment by whitlockjc** [comment]
PRs are always welcome.  I am currently working on a json-refs release and plan on updating swagger-tools and sway to include the new version.  So I'll likely wait until that's done.

**Comment by pixelshaded** [comment]
https://github.com/apigee-127/swagger-tools/pull/402
